File: src/main/java/kr/mjc/rachel/concepts/functions/FunctionExample.java
/*
 * @FunctionalInterface : abstract 메서드가 1개 있는 인터페이스
 * Function<T,R> : 인자가 1개 있고, 리턴값이 있는 함수를 갖는 @FunctionalInterface
 */
package kr.mjc.rachel.concepts.functions;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class FunctionExample {
  /**
   * list의 요소에 function.apply(t)를 실행해서 새로운 리스트를 만든다.
   */
  public <T, R> List<R> transform(List<T> list, Function<T, R> function) {
    List<R> result = new ArrayList<>();
    for (T t : list) result.add(function.apply(t));
    return result;
  }

  public static void main(String[] args) {
    FunctionExample exam = new FunctionExample();
    List<Integer> list1 = List.of(1, 2, 3);

    Function<Integer, Integer> func = new Function<>() {
      @Override
      public Integer apply(Integer n) {
        return n * 10;
      }
    };

    List<Integer> list2 = exam.transform(list1, func);
    System.out.println(list2);
  }
}

File: src/main/kotlin/kr/mjc/jacob/concepts/functions/FunctionExample.kt
package kr.mjc.jacob.concepts.functions

class FunctionExample {
  // higher order function : 함수를 인자로 받거나 함수를 리턴하는 함수
  // inline 함수 인스턴스를 만들지 않고 함수의 코드를 삽입한다.
  inline fun <T, R> transform(list: List<T>, apply: (T) -> R): List<R> {
    val result = mutableListOf<R>()
    for (e in list) result.add(apply(e))
    return result
  }

  fun tenTimes(n: Int): Int = n * 10
}

fun main() {
  val exam = FunctionExample()
  val list1 = listOf(1, 2, 3)

  // lamda expression
  val f1: (Int) -> Int = { n -> n * 10 }
  val list2 = exam.transform(list1, f1)
  val list3 = exam.transform(list1) { n -> n * 10 }
  val list4 = exam.transform(list1, exam::tenTimes) // method reference

  println("list2 = $list2")
  println("list3 = $list3")
  println("list4 = $list4")

  // collection.map 은 transform 함수
  val list5 = list1.map { n -> n * 10 }
  val list6 = list1.map(exam::tenTimes)
  println("list5 = $list5")
  println("list6 = $list6")
}
